server/apps/install_app.go

```diff
@@ -5,40 +5,120 @@ package apps
 
 import (
 	"errors"
+	"fmt"
+	"net/http"
 
 	"github.com/mattermost/mattermost-plugin-apps/server/utils/md"
+	"github.com/mattermost/mattermost-server/v5/model"
 )
 
 type InInstallApp struct {
 	ActingMattermostUserID string
 	NoUserConsentForOAuth2 bool
 	Manifest               *Manifest
 	Secret                 string
+	SessionToken           string
 }
 
 type OutInstallApp struct {
 	md.MD
 	App *App
 }
 
+type InstallRequest struct {
+	OAuthAPIKey string
+	OAuthSecret string
+}
+
 func (r *registry) InstallApp(in *InInstallApp) (*OutInstallApp, error) {
 	if in.Manifest.AppID == "" {
 		return nil, errors.New("app ID must not be empty")
 	}
 	// TODO check if acting user is a sysadmin
 
 	// TODO remove mock, implement for real
+
+	bot, token, err := r.createBot(in.Manifest, in.SessionToken)
+	if err != nil {
+		return nil, err
+	}
+	oAuthApp, err := r.createOAuthApp(in.ActingMattermostUserID, token.Token, in.Manifest)
+	if err != nil {
+		return nil, err
+	}
+
 	app := &App{
 		Manifest:               in.Manifest,
 		GrantedPermissions:     in.Manifest.RequestedPermissions,
 		NoUserConsentForOAuth2: in.NoUserConsentForOAuth2,
 		Secret:                 in.Secret,
+		BotID:                  bot.UserId,
+		BotToken:               token.Token,
+		OAuthAppID:             oAuthApp.Id,
+		OAuthSecret:            oAuthApp.ClientSecret,
 	}
 	r.apps[in.Manifest.AppID] = app
 
+	extraInfo := fmt.Sprintf(`Bot token: %s
+	OAuth Client ID: %s
+	OAuth Client Secret: %s`, token.Token, oAuthApp.Id, oAuthApp.ClientSecret)
+
 	out := &OutInstallApp{
-		MD:  md.Markdownf("Installed %s (%s)", in.Manifest.DisplayName, in.Manifest.AppID),
+		MD:  md.Markdownf("Installed %s (%s)\n%s", in.Manifest.DisplayName, in.Manifest.AppID, extraInfo),
 		App: app,
 	}
 	return out, nil
 }
+
+func (r *registry) createBot(manifest *Manifest, sessionToken string) (*model.Bot, *model.UserAccessToken, error) {
+	client := model.NewAPIv4Client(r.configurator.GetConfig().MattermostSiteURL)
```

Comment:
All APIv4Client on this file will not work because there is no logged user to the client. I am not sure how to get an authenticated client in this context.

Reply:
0/5 seems the most logical choice is to either create the new needed p.API methods, or to find a way to create a token from the sessionID that we have.

Reply:
More confused... We have a `p.API.GetSession` and it has the token - why not use it?

Reply:
Maybe I am missing something. This is done we hit a button in a modal dialog. Therefore, this comes from the HTTP handler. The HTTP handler only have access to the request and the response writer, neither of which contains the session ID nor anything similar.

The base ServeHTTP function does have the plugin.Context object that contains the sessionID, but it is "lost" when calling the ServeHTTP of the router. I will dig deeper into how to pass this information (maybe adding an extra header in the request?).

Reply:
:+1